saleor/attribute/models/base.py

```diff
@@ -214,6 +214,120 @@ def get_translated_keys(self):
+        query = self._prepare_query_for_bulk_operation(objects_data)
+
+        # iterate over all records in db and check if they match any of objects data
+        for record in query.iterator():
+            # iterate over all objects data and check if they match any of records in db
+            for index, obj in objects_enumerated:
+                if self._is_correct_record(record, obj):
+                    # upon finding existing record add it to results
+                    results.append((index, record))
+                    # remove it from objects list, so it won't be added to new records
+                    objects_enumerated.remove((index, obj))
+
+                    break
+
+        # add what is left to the list of new records
+        self._add_new_records(objects_enumerated, objects_not_in_db, results)
+        # sort results by index as db record order might be different from sort_order
+        results.sort()
+        results = [obj for index, obj in results]
+
+        if objects_not_in_db:
+            self.bulk_create(objects_not_in_db)  # type: ignore[arg-type]
+
+        return results
+
+    def bulk_update_or_create(self, objects_data):
+        # this method mimics django's queryset.update_or_create method on bulk objects
+        # https://docs.djangoproject.com/en/5.0/ref/models/querysets/#update-or-create
+        results = []
+        objects_not_in_db: List[AttributeValue] = []
+        objects_to_be_updated = []
+        update_fields = set()
+        objects_enumerated = list(enumerate(objects_data))
+        query = self._prepare_query_for_bulk_operation(objects_data)
+
+        # iterate over all records in db and check if they match any of objects data
+        for record in query.iterator():
+            # iterate over all objects data and check if they match any of records in db
+            for index, obj in objects_enumerated:
+                if self._is_correct_record(record, obj):
+                    # upon finding a matching record, update it with defaults
+                    for key, value in obj["defaults"].items():
+                        setattr(record, key, value)
+                        update_fields.add(key)
+
+                    # add it to results and objects to be updated
+                    results.append((index, record))
+
+                    # add it to objects to be updated, so it can be bulk updated later
+                    objects_to_be_updated.append(record)
+
+                    # remove it from objects data, so it won't be added to new records
+                    objects_enumerated.remove((index, obj))
+
+                    break
+
+        # add what is left to the list of new records
+        self._add_new_records(objects_enumerated, objects_not_in_db, results)
+
+        # sort results by index as db record order might be different from sort_order
+        results.sort()
+        results = [obj for index, obj in results]
+
+        if objects_not_in_db:
```

Comment:
How do you protect against thread races?
For me, the list built in lines 282-300 could be outdated here as other threads could apply changes to the objects.